sys/mutex.c

```diff
@@ -40,6 +45,13 @@ void mtx_unlock(mtx_t *m) {
   WITH_NO_PREEMPTION {
     m->m_owner = NULL;
     m->m_lockpt = NULL;
-    sleepq_signal(&m->m_owner);
+    turnstile_chain_lock(m);
+    turnstile_t *ts = turnstile_lookup(m);
+    if (ts != NULL) {
+      // TODO signal instead of broadcast (not implemented yet)
+      turnstile_broadcast(ts);
```

Comment:
@ilikeheaps 
In "The Design and Implementation of the FreeBSD Operating System" they say:

> The lower cost for releasing an uncontested lock explains the algorithm used to awaken waiters on a mutex. Historically, the mutex code would only awaken a single waiter when a contested lock was released, which left the lock in a contested state if there were more than one waiter. However, leaving a contested lock ensured that the new lock holder would have to perform a more expensive unlock operation. Indeed, all but the last waiter would have an expensive unlock operation. In the current FreeBSD system, all the waiters are awakened when the lock is released. Usually they end up being scheduled sequentially, which results in them all getting to do cheaper unlock operations. If they do all end up running concurrently, they will then use adaptive spinning and will finish the chain of lock requests sooner since the context switches to awaken the threads are performed in parallel rather than sequentially. This change in behavior was motivated by documentation of these effects noted in Solaris Internals [McDougall & Mauro, 2006].

So maybe we don't want signal instead of broadcast. But I'm not sure.

Reply:
I agree. I'll change the comment appropriately.

Reply:
I've added location of this explanation in the book to your comment.